src/map/skill.cpp

```diff
@@ -9499,36 +9499,25 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 	case NPC_TRANSFORMATION:
 	case NPC_METAMORPHOSIS:
 		if(md && md->skill_idx >= 0) {
-			int class_ = mob_random_class (md->db->skill[md->skill_idx]->val,0);
+			int class_ = mob_random_class (md->db->skill[md->skill_idx]->summons);
 			if (skill_lv > 1) //Multiply the rest of mobs. [Skotlex]
-				mob_summonslave(md,md->db->skill[md->skill_idx]->val,skill_lv-1,skill_id);
+				mob_summonslave(md,md->db->skill[md->skill_idx]->summons,skill_lv-1,skill_id);
 			if (class_) mob_class_change(md, class_);
 		}
 		break;
 
 	case NPC_EMOTION_ON:
-	case NPC_EMOTION:
-		//val[0] is the emotion to use.
 		//NPC_EMOTION & NPC_EMOTION_ON can change a mob's mode 'permanently' [Skotlex]
-		//val[1] 'sets' the mode
-		//val[2] adds to the current mode
-		//val[3] removes from the current mode
-		//val[4] if set, asks to delete the previous mode change.
+		// NPC_EMOTION_ON adds the mode to the current mode
 		if(md && md->skill_idx >= 0 && tsc)
 		{
-			clif_emotion(bl, md->db->skill[md->skill_idx]->val[0]);
-			if(md->db->skill[md->skill_idx]->val[4] && tsce)
-				status_change_end(bl, type, INVALID_TIMER);
-
-			//If mode gets set by NPC_EMOTION then the target should be reset [Playtester]
-			if(!battle_config.npc_emotion_behavior && skill_id == NPC_EMOTION && md->db->skill[md->skill_idx]->val[1])
-				mob_unlocktarget(md,tick);
+			int mode_passive = (md->db->skill[md->skill_idx]->mob_mode & MD_AGGRESSIVE) ? 0 : MD_AGGRESSIVE;	// Remove aggressive mode when the new mob type is passive.
 
-			if(md->db->skill[md->skill_idx]->val[1] || md->db->skill[md->skill_idx]->val[2])
-				sc_start4(src,src, type, 100, skill_lv,
-					md->db->skill[md->skill_idx]->val[1],
-					md->db->skill[md->skill_idx]->val[2],
-					md->db->skill[md->skill_idx]->val[3],
+			if (md->db->skill[md->skill_idx]->mob_mode > -1)
```

Comment:
Looks like you gave NPC_EMOTION_ON a new functionality. Can you please document your detailed tests on which base you are changing this and link it to this PR? I couldn't find an issue that's being linked to this.

If needed, I can help checking on pre-renewal to make sure this isn't a renewal-only change.

Reply:
NPC_EMOTION_ON and NPC_EMOTION, after this commit, should work exactly as before. Both skills did nothing when no mode was provided. This commit just extends the functionality to set the AI of the mob when it casts another skill :
```
// Monster Name,    State,    Skill ID,    Level,    Probability,    Cast Time,    Delay Time,    NO_DISPEL,    AI Setting,    SEND_EMOTICON/SEND_CHAT,     Emotion #/Chat #,    IF,    Condition value
E_NYAIL_PIG    BERSERK_ST    AL_DECAGI    5    700    500    15000    NO_DISPEL    MONSTER_TYPE_17
E_SALAMANDER    RUSH_ST    WZ_METEOR    9    300    500    20000        MONSTER_TYPE_21
E_SALAMANDER    RUSH_ST    NPC_WIDESILENCE    5    500    700    30000        MONSTER_TYPE_21    SEND_EMOTICON    9
E_SALAMANDER    RUSH_ST    NPC_ALLHEAL    1    1000    10000    30000    NO_DISPEL    MONSTER_TYPE_21    SEND_EMOTICON    28    IF_HP    10
GRIM_REAPER_ANKOU    BERSERK_ST    NPC_COMET    3    100    10000    75000        MONSTER_TYPE_21
GRIM_REAPER_ANKOU    BERSERK_ST    NPC_DARKCROSS    10    50    2000    12000        MONSTER_TYPE_21
GRIM_REAPER_ANKOU    BERSERK_ST    NPC_DRAGONFEAR    5    100    3000    30000        MONSTER_TYPE_21
TENDRILRION    BERSERK_ST    PF_SPIDERWEB    1    300    0    10000        MONSTER_TYPE_04    SEND_EMOTICON    7
```

The lines are differents here because this commit no longer uses the val[] set when loading mob_skill_db :
```
// from mob_parse_row_mobskilldb (txt format)
if(ms->skill_id == NPC_EMOTION && mob_id > 0 &&
		ms->val[1] == mob->status.mode)
	{
		ms->val[1] = 0;
		ms->val[4] = 1; //request to return mode to normal.
	}
	if(ms->skill_id == NPC_EMOTION_ON && mob_id > 0 && ms->val[1])
	{	//Adds a mode to the mob.
		//Remove aggressive mode when the new mob type is passive.
		if (!(ms->val[1]&MD_AGGRESSIVE))
			ms->val[3] |= MD_AGGRESSIVE;
		ms->val[2] |= ms->val[1]; //Add the new mode.
		ms->val[1] = 0; //Do not "set" it.
```

Basically from what I understand from the current code, NPC_EMOTION 'sets' (or removes) a mode depending of the mode provided while NPC_EMOTION_ON 'adds' the mode provided to the current mode.

Reply:
I always thought the "adding" of a mode was just a custom feature, especially since Aegis doesn't even have those granular modes that we have. It didn't depend on NPC_EMOTION vs. NPC_EMOTION_ON, but rather on whether you set var2, var3 and/or var4.
In our default mob_skill_db, only var1 was ever used AFAIK.

We probably need more tests on NPC_EMOTION_ON.